Add functionality to update help in user scope

adityapatwardhan opened the pull request:
Fixes #2190 

## PR Summary

Update-Help needs elevation to update the help content.
This PR adds a functionality to save the updated help content under $HOME to elevation is not required.
To enable this functionality -Scope parameter is added Update-Help.
Scope is defaulted to CurrentUser.
CurrentUser scope saves the help content under $HOME/PowerShell/Help.
Also made changes to Get-Help to lookup help from under $HOME.
Updated and added tests for Update-Help and Get-Help.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [X] Issue filed - Issue link: https://github.com/PowerShell/PowerShell-Docs/issues/2168
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [X] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

adityapatwardhan committed:
[Feature] Add functionality to update help in user scope

Update-Help needs elevation to update the help content.
This PR adds a functionality to save the updated help content under $HOME to elevation is not required.
To enable this functionality -Scope parameter is added Update-Help.
Scope is defaulted to Global, which adheres the current behavior, whereas User scope saves the help content under $HOME.
Also made changes to Get-Help to lookup help from under $HOME.
Updated and added tests for Update-Help and Get-Help.

daxian-dbw commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -530,6 +539,44 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
             string location = MUIFileSearcher.LocateFile(helpFileToLoad, searchPaths);
 
+            // try home location
+            if(String.IsNullOrEmpty(location))
```
I believe the code here can be integrated into the `if .. else if ... else` block in the `if (!File.Exists(helpFileToLoad))`.

The code looks like this:
```c#
if (!File.Exists(helpFileToLoad))
{
    helpFileToLoad = Path.GetFileName(helpFileToLoad);

    if (mshSnapInInfo != null)
    {
        Diagnostics.Assert(!string.IsNullOrEmpty(mshSnapInInfo.ApplicationBase),
            "Application Base is null or empty.");
        // not minishell case..
        // we have to search only in the application base for a mshsnapin...
        // if you create an absolute path for helpfile, then MUIFileSearcher
        // will look only in that path.
        searchPaths.Add(mshSnapInInfo.ApplicationBase);
        searchPaths.Add(HelpUtils.GetUserHomeHelpSearchPath());
    }
    else if (cmdletInfo.Module != null && !string.IsNullOrEmpty(cmdletInfo.Module.Path))
    {
        searchPaths.Add(cmdletInfo.Module.ModuleBase);
        // calulate the module folder path in UserHomeHelpSearchPath with the code below.
        searchPaths.Add(calulatedUserSearchModulePath);
    }
    else
    {
        searchPaths.Add(GetDefaultShellSearchPath());
        searchPaths.Add(GetCmdletAssemblyPath(cmdletInfo));
        searchPaths.Add(HelpUtils.GetUserHomeHelpSearchPath());
    }
}
else
{
    helpFileToLoad = Path.GetFullPath(helpFileToLoad);
}

string location = MUIFileSearcher.LocateFile(helpFileToLoad, searchPaths);
```

The current code passes in `helpFileToLoad` in case that `cmdletInfo.mshSnapInInfo` or `cmdletInfo.module` is available. This is not necessary because if you look at `MUIFileSearcher.LocateFile`, even if `helpFileToLoad` is a full file path, it still gets first normalized to `filename` and `parent folder path`, and then search the `filename` in the folders.

adityapatwardhan replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -0,0 +1,32 @@
+        /// <summary>
+        /// Get the path to $HOME
+        /// </summary>
+        internal static string GetUserHomeHelpSearchPath()
```
This could be made a static property with a static field that caches the path.

adityapatwardhan replied:
Fixed.

kalgiz commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -521,6 +523,13 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
+                    // in case of Microsoft.PowerShell.Core module, cmdletInfo.ModuleName is null.
+                    // if cmdletInfo.ModuleName is null, Path.Combine return HelpUtils.GetUserHomeHelpSearchPath()
+                    // So we will be searching under HelpUtils.GetUserHomeHelpSearchPath()
+                    // The search is done in order of the searchPath list. We keep the userHelp
```
nit: you can use the capital letters at the beginning of the sentence and punctuation:

> // In case of Microsoft.PowerShell.Core module, cmdletInfo.ModuleName is null.
> // If cmdletInfo.ModuleName is null, Path.Combine returns HelpUtils.GetUserHomeHelpSearchPath(),
> // so we will be searching under HelpUtils.GetUserHomeHelpSearchPath().
> // The search is done in order of the searchPath list.

adityapatwardhan replied:
Fixed.

Francisco-Gamino commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -0,0 +1,32 @@
+                homeFolder = Environment.GetEnvironmentVariable("HOME");
+            }
+
+            var homeHelpFolder = Path.Combine(homeFolder, "PowerShellHelp");
```
This is a very good suggestion.

adityapatwardhan replied:
Fixed.

adityapatwardhan commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -318,6 +318,8 @@ private HelpInfo GetHelpInfo(CommandInfo commandInfo, bool reportErrors, bool se
                 GetModulePaths(commandInfo, out moduleName, out moduleDir, out nestedModulePath);
 
                 Collection<String> searchPaths = new Collection<String>();
+                searchPaths.Add(HelpUtils.GetUserHomeHelpSearchPath());
```
Fixed.

adityapatwardhan commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -530,6 +539,44 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
+
+                    if (moduleBase.EndsWith(moduleName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        homeModulePath = Path.Combine(homeHelpPath, cmdletInfo.ModuleName);
```
Fixed.

adityapatwardhan commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -530,6 +539,44 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
+                    else
+                    {
+                        // This module has version folder.
+                        var moduleVersion = (new DirectoryInfo(moduleBase)).Name;
```
Fixed.

adityapatwardhan commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -530,6 +539,44 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
+                    {
+                        // This module has version folder.
+                        var moduleVersion = (new DirectoryInfo(moduleBase)).Name;
+                        homeModulePath = Path.Combine(homeHelpPath,cmdletInfo.ModuleName, moduleVersion);
```
Fixed.

adityapatwardhan commented on src/System.Management.Automation/help/UpdatableHelpCommandBase.cs:
```diff
@@ -115,6 +115,16 @@ public SwitchParameter Force
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true)]
+        public UpdateHelpScope Scope
+        {
+            internal get;
```
Fixed.

adityapatwardhan commented on src/System.Management.Automation/help/UpdatableHelpCommandBase.cs:
```diff
@@ -862,4 +872,19 @@ internal void ProcessException(string moduleName, string culture, Exception e)
+        /// <summary>
+        /// Save the help content to the module directory. This is the default behavior.
+        /// </summary>
+        Global,
```
Fixed.

adityapatwardhan commented on src/System.Management.Automation/help/UpdateHelpCommand.cs:
```diff
@@ -375,7 +375,42 @@ internal override bool ProcessModuleWithCulture(UpdatableHelpModuleInfo module,
+
+                            // In case of inbox modules, the help is put under $PSHOME/<current_culture>,
+                            // since the dlls are not published under individual module folders, but under $PSHome
+                            // In case of other modules, the help in under module.ModuleBase/<current_culture>
```
Fixed.

adityapatwardhan commented on src/System.Management.Automation/help/UpdateHelpCommand.cs:
```diff
@@ -375,7 +375,42 @@ internal override bool ProcessModuleWithCulture(UpdatableHelpModuleInfo module,
+
+                            string parentModuleBase = Directory.GetParent(module.ModuleBase).Name;
+
+                            // In case of inbox modules, the help is put under $PSHOME/<current_culture>,
```
Yes, this is explaining the various locations the Allusers help can be put at and the code below adjusts that for CurrentUser scope. Improved the comment.

Francisco-Gamino commented on src/System.Management.Automation/help/UpdateHelpCommand.cs:
```diff
@@ -375,7 +375,42 @@ internal override bool ProcessModuleWithCulture(UpdatableHelpModuleInfo module,
                         // Gather destination paths
                         Collection<string> destPaths = new Collection<string>();
 
-                        destPaths.Add(module.ModuleBase);
```
The logic at line 350 might have to be revised.
if (Utils.IsUnderProductFolder(module.ModuleBase) && (!Utils.IsAdministrator()))
{
    string message = StringUtil.Format(HelpErrors.UpdatableHelpRequiresElevation);
    ProcessException(module.ModuleName, null, new UpdatableHelpSystemException("UpdatableHelpSystemRequiresElevation",
            message, ErrorCategory.InvalidOperation, null, null));
    return false;
}

adityapatwardhan replied:
Fixed.

Francisco-Gamino replied:
@adityapatwardhan : As per our conversation, we need to update UpdatableHelpRequiresElevation to include the UpdateHelpScope change.

adityapatwardhan commented on src/System.Management.Automation/help/UpdateHelpCommand.cs:
```diff
@@ -375,7 +375,42 @@ internal override bool ProcessModuleWithCulture(UpdatableHelpModuleInfo module,
+                            // In case of inbox modules, the help is put under $PSHOME/<current_culture>,
+                            // since the dlls are not published under individual module folders, but under $PSHome
+                            // In case of other modules, the help in under module.ModuleBase/<current_culture>
+                            if(module.ModuleBase.EndsWith(module.ModuleName, StringComparison.OrdinalIgnoreCase))
```
That is correct, the `homeFolder` is the root for the help content. A locale folder will be created under it for saving the help.

adityapatwardhan committed:
[Feature] Address code review feedback

adityapatwardhan commented:
@daxian-dbw @Francisco-Gamino @kalgiz Code review comments have been addressed. Please have a look.

adityapatwardhan commented:
I have found an issue regarding writing the HelpInfo.xml. Marking this PR as WIP.

adityapatwardhan committed:
[Feature] Fixed issue with HelpInfo xml

We need to download the helpinfo xml in case of updatable version or if HelpInfo was never present.
Made updates so when specifying CurrentUser scope, HelpInfo is downloaded to the user help folder.

adityapatwardhan commented:
@daxian-dbw @Francisco-Gamino I have fixed the issues with HelpInfo xml. Please have another look.

daxian-dbw commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -0,0 +1,74 @@
+        /// </summary>
+        internal static string GetUserHomeHelpSearchPath()
+        {
+            if (userScopeRootPath == null)
```
I don't see `userScopeRootPath` gets used except for calculating `userHomeHelpPath`. Maybe it shouldn't be made a static variable.

adityapatwardhan replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -540,6 +551,41 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
+
+            // In case of inbox modules, the help is put under $PSHOME/<current_culture>,
+            // since the dlls are not published under individual module folders, but under $PSHome
+            // In case of other modules, the help in under module.ModuleBase/<current_culture>
```
> the help in under

**is** under?

adityapatwardhan replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -511,16 +512,26 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
                     searchPaths.Add(GetDefaultShellSearchPath());
                     searchPaths.Add(GetCmdletAssemblyPath(cmdletInfo));
+
+                    // In case of Microsoft.PowerShell.Core module, cmdletInfo.ModuleName is null.
```
I don't understand the comment here. For cmdlets from `Microsoft.PowerShell.Core`, `mshSnapInInfo` won't be `null` and thus will be covered by the first if block.

adityapatwardhan replied:
Removed the comment. The code was refactored in to helper in HelpUtils.cs. The comment was left behind.

daxian-dbw commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -511,16 +512,26 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
+                    // If cmdletInfo.ModuleName is null, Path.Combine returns HelpUtils.GetUserHomeHelpSearchPath()
+                    // So we will be searching under HelpUtils.GetUserHomeHelpSearchPath()
+                    // The search is done in order of the searchPath list.
+                    var homeModulePath = Path.Combine(HelpUtils.GetUserHomeHelpSearchPath(), cmdletInfo.ModuleName);
```
I believe `cmdletInfo.ModuleName` will be null when the code reaches here.

daxian-dbw commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -540,6 +551,41 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
             return location;
         }
 
+        private string CalculateHelpSearchPath(CmdletInfo cmdletInfo)
```
I don't see this method used anywhere.

adityapatwardhan replied:
Good catch, got refactored into HelpUtils.cs

daxian-dbw commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -0,0 +1,74 @@
+            // If the the scope is AllUsers, then the help goes under moduleBase.
+
+            var userHelpPath = GetUserHomeHelpSearchPath();
+            string parentModuleBase = Directory.GetParent(moduleBase).Name;
```
> parentModuleBase 

Maybe renamed to `moduleBaseParent`

adityapatwardhan replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -0,0 +1,74 @@
+                //This module is not an inbox module, so help goes under <userHelpPath>/<moduleName>
+                newModuleBase = Path.Combine(userHelpPath, moduleName);
+            }
+            else if (String.Compare(parentModuleBase, moduleName, StringComparison.OrdinalIgnoreCase) == 0)
```
nit: `String.Compare` can be replaced with `String.Equals(a,b,StringComparison)`.

adityapatwardhan replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/help/UpdateHelpCommand.cs:
```diff
@@ -375,7 +382,12 @@ internal override bool ProcessModuleWithCulture(UpdatableHelpModuleInfo module,
+                            Directory.CreateDirectory(moduleBase);                        
+                        }
+                        
+                        destPaths.Add(moduleBase);                        
```
There are trailing spaces in this line.

adityapatwardhan replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/help/UpdateHelpCommand.cs:
```diff
@@ -375,7 +382,12 @@ internal override bool ProcessModuleWithCulture(UpdatableHelpModuleInfo module,
-                        destPaths.Add(module.ModuleBase);
+                        if(!Directory.Exists(moduleBase))
+                        {
+                            Directory.CreateDirectory(moduleBase);                        
```
Trailing spaces should be removed.

adityapatwardhan replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/help/UpdateHelpCommand.cs:
```diff
@@ -347,7 +354,7 @@ internal override bool ProcessModuleWithCulture(UpdatableHelpModuleInfo module,
                 }
 
-                if (Utils.IsUnderProductFolder(module.ModuleBase) && (!Utils.IsAdministrator()))
+                if (Utils.IsUnderProductFolder(moduleBase) && (!Utils.IsAdministrator()) && (this.Scope == UpdateHelpScope.AllUsers))
```
> && (this.Scope == UpdateHelpScope.AllUsers)

Is this really needed?

adityapatwardhan replied:
Yes, so we only throw an error that elevation is required if we are updating AllUsers scope.

daxian-dbw replied:
The `moduleBase` here won't be under product folder in case that `scope == currentuser`, so no error will be thrown out when `scope == currentuser`.

adityapatwardhan replied:
Fixed.

adityapatwardhan committed:
[Feature] Address code review feedback

Removed unnecessary check for UpdateHelpScope

Address code review feedback

* Changed default value for scope to be CurrentUser.
* Updated tests.

adityapatwardhan commented:
@daxian-dbw @Francisco-Gamino code review feedback has been addressed.

adityapatwardhan commented:
@daxian-dbw @Francisco-Gamino all comments addressed.

daxian-dbw commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -511,16 +512,22 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
                     searchPaths.Add(GetDefaultShellSearchPath());
                     searchPaths.Add(GetCmdletAssemblyPath(cmdletInfo));
+
+                    var homeModulePath = Path.Combine(HelpUtils.GetUserHomeHelpSearchPath(), cmdletInfo.ModuleName);
```
> cmdletInfo.ModuleName

I believe `cmdletInfo.ModuleName` will be null when the code reaches here.
Maybe we should only add `HelpUtils.GetUserHomeHelpSearchPath()`? Just like the `GetDefaultShellSearchPath`,

adityapatwardhan replied:
Sorry missed that in the last update. Fixed now.

adityapatwardhan commented:
@Francisco-Gamino Can you have another look?

Francisco-Gamino commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -0,0 +1,69 @@
+    internal class HelpUtils
+    {
+        private static string userHomeHelpPath = null;
+
```
np: Please remove the extra space.

adityapatwardhan replied:
Fixed.

SteveL-MSFT commented on src/System.Management.Automation/help/CommandHelpProvider.cs:
```diff
@@ -511,16 +512,20 @@ private string GetHelpFile(string helpFile, CmdletInfo cmdletInfo)
                 {
                     searchPaths.Add(GetDefaultShellSearchPath());
                     searchPaths.Add(GetCmdletAssemblyPath(cmdletInfo));
+                    searchPaths.Add(HelpUtils.GetUserHomeHelpSearchPath());
```
Does the ordering matter here?  If it finds help in the DefaultShellSearchPath, does it stop searching further?  If so, I think the user home help path should be first as that's most likely to be up-to-date

adityapatwardhan replied:
Updated to give preference to user scope path.

SteveL-MSFT commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -0,0 +1,69 @@
+#else
+                string userScopeRootPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), "PowerShell");
+#endif
+                userHomeHelpPath = Path.Combine(userScopeRootPath, "Help");
```
Do we need to worry about SxS PSCore6.x?

adityapatwardhan replied:
Fixed. Added a PSVersion string in path.

adityapatwardhan replied:
Reverted this change, since the source location for help can only host 1 help content version. So even if we have SxS PowerShell versions, the content only has 1 published version.

SteveL-MSFT commented on src/System.Management.Automation/help/UpdatableHelpCommandBase.cs:
```diff
@@ -862,4 +872,19 @@ internal void ProcessException(string moduleName, string culture, Exception e)
+    public enum UpdateHelpScope
+    {
+        /// <summary>
+        /// Save the help content to the module directory. This is the default behavior.
```
Personally, I don't like that we default to requiring elevation as it means that for most users they see an error and then need to rerun `Update-Help -Scope CurrentUser`.  Let's have this reviewed by @PowerShell/powershell-committee as it would be a breaking change.

adityapatwardhan replied:
Updated.

joeyaiello commented:
Forgot to post my logic on this one: 

I love it. Been hoping for this for a very long time. I think Jeffrey even filed an issue before alpha went public. 

I'm totally okay with this as a breaking change. Anyone with an advanced enough scenario to take a dependency on the path of the XML, they can figure out what's happening. To be honest, it's most likely to break our own tests in PowerShell-Docs than anything else. 

I think we should throw a warning for a version or two though. I know that we usually say warnings are useless for non interactive scenarios because they're headless, but I think we're getting used more and more in CI and orchestration scenarios where people can go review the console output afterwards. A yellow warning on Appveyor, e.g. would be very useful.

adityapatwardhan commented:
@joeyaiello I can make the changes for updating the default. I would prefer not to add a warning message but, instead get feedback from community if they do not like the new default.

Once this is merged I will make the necessary changes on PowerShell-Docs repo.

adityapatwardhan commented:
@SteveL-MSFT Your feedback has been addressed. Please have another look.

adityapatwardhan committed:
[Feature] Fix tab completion test

[Feature] Improve test title

[Feature] Fix tests and remove version from path

daxian-dbw commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -12,21 +12,21 @@ internal class HelpUtils
                 string userScopeRootPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), "PowerShell");
 #endif
-                userHomeHelpPath = Path.Combine(userScopeRootPath, "Help");
+                userHomeHelpPath = Path.Combine(userScopeRootPath, "Help", psVersion);
```
The name may be too specific, so my 6.0.3 PSCore cannot use the existing help content I downloaded with 6.0.1 ...

daxian-dbw commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -0,0 +1,69 @@
+#else
+                string userScopeRootPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), "PowerShell");
+#endif
+                userHomeHelpPath = Path.Combine(userScopeRootPath, "Help", psVersion);
```
The name may be too specific, so my 6.0.3 PSCore cannot use the existing help content I downloaded with 6.0.1 ...

adityapatwardhan replied:
Changed to remove version from path.

SteveL-MSFT commented on test/powershell/engine/Help/HelpSystem.Tests.ps1:
```diff
@@ -31,7 +31,7 @@ function RunTestCase
     $moduleName = "Microsoft.PowerShell.Core"
 
-    UpdateHelpFromLocalContentPath $moduleName $tag
+    UpdateHelpFromLocalContentPath $moduleName $tag -Scope 'AllUsers'
```
Does this need to be `AllUsers`?  Seems like most of the help tests can validate against CurrentUser and we should have a few tests explicitly validating `AllUsers`

adityapatwardhan replied:
Fixed.

SteveL-MSFT commented on test/powershell/engine/Help/HelpSystem.Tests.ps1:
```diff
@@ -326,3 +326,59 @@ Describe "help function uses full view by default" -Tags "CI" {
+        ## Delete help from global scope if it exists.
+        $currentCulture = (Get-Culture).Name
+
+        $managementHelpFilePath = Join-Path $PSHOME -ChildPath $currentCulture -AdditionalChildPath 'Microsoft.PowerShell.Commands.Management.dll-Help.xml'
```
LInes 352-367 very similar, perhaps change it to a `foreach` loop?

adityapatwardhan replied:
Refactored the test file. There is some duplication, but it is much simpler to read now.

SteveL-MSFT commented on test/powershell/engine/Help/UpdatableHelpSystem.Tests.ps1:
```diff
@@ -151,10 +173,21 @@ function ValidateInstalledHelpContent
     )
 
-    $helpFilesInstalled = @(GetFiles -path $testCases[$moduleName].HelpInstallationPath | ForEach-Object {Split-Path $_ -Leaf})
+    if($Scope)
```
Shouldn't this be `$UserScope`?

SteveL-MSFT commented on test/powershell/engine/Help/UpdatableHelpSystem.Tests.ps1:
```diff
@@ -168,44 +201,71 @@ function RunUpdateHelpTests
         [string]$tag = "CI",
-        [switch]$useSourcePath
+        [switch]$useSourcePath,
+        [switch]$Scope
```
`$UserScope` is more descriptive than just `$Scope` for a switch and is consistent with previous use

adityapatwardhan replied:
Fixed.

adityapatwardhan commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -947,7 +947,7 @@ Describe "Tab completion help test" -Tags @('RequireAdminOnWindows', 'CI') {
         if (-not (Test-Path (Join-Path $aboutHelpPath "about_Splatting.help.txt")))
         {
-            Update-Help -Force -ErrorAction SilentlyContinue
+            Update-Help -Force -ErrorAction SilentlyContinue -Scope 'AllUsers'
```
Fixed.

adityapatwardhan committed:
[Feature] Test fixes and updated to CompletionCompletors and HelpFileHelpProvider

* Updated tab completion logic to look up about help topics from user path as well.
* Updated HelpFileHelpProvider to look up help from user path.
* Added deduplication logic in HelpFileHelpProvider.
* Refactored Get-Help tests to get help from local source only.
* Refactored Get-Help tests to have more tests in user scope and a few in all users scope.
* Updated assets as they were missing about help files.
* Addressed code review feedback.

adityapatwardhan committed:
[Feature] Remove unnecessary whitespace

adityapatwardhan commented:
@SteveL-MSFT  Addressed all comments. Refactored tests. Added new help content. Fixed issues with tabcompletion and fixed cab file read/write permission usage.

daxian-dbw commented on src/System.Management.Automation/help/HelpFileHelpProvider.cs:
```diff
@@ -321,17 +343,11 @@ internal Collection<string> GetExtendedSearchPaths()
-                searchPaths.Insert(0, defaultShellSearchPath);
-            }
+            // Add $PSHome path. This ensures that default.help.txt is picked before other help content for 'Get-Help'
+            searchPaths.Add(GetDefaultShellSearchPath());
```
The path added here won't be at the index 0, please revisit.

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -5919,7 +5919,21 @@ private static string GetNamespaceToRemove(CompletionContext context, TypeComple
+            var directorySeparatorChar = Path.DirectorySeparatorChar;
+
+            // Add the user scope path first, since it is searched in order.
+            var userHelpRoot = HelpUtils.GetUserHomeHelpSearchPath() + directorySeparatorChar + currentCulture;
```
Use Path.Combine

adityapatwardhan replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/help/CabinetNativeApi.cs:
```diff
@@ -225,9 +225,16 @@ internal static void FdiFree(IntPtr memblock)
             FileAccess access = CabinetNativeApi.ConvertPermissionModeToFileAccess(pmode);
             FileShare share = CabinetNativeApi.ConvertPermissionModeToFileShare(pmode);
 
+            if(mode == FileMode.Open || mode == FileMode.OpenOrCreate)
```
>e.Open || [](start = 30, length = 9)

Please add comments

adityapatwardhan replied:
Fixed.

adityapatwardhan committed:
[Feature] Addressed code review comments

adityapatwardhan commented:
@daxian-dbw This is ready for merge.

daxian-dbw commented:
@adityapatwardhan Please update the PR summary since some design changes were made to this PR.

iSazonov commented on test/powershell/engine/Help/UpdatableHelpSystem.Tests.ps1:
```diff
@@ -308,7 +380,19 @@ Describe "Validate Update-Help from the Web for all PowerShell Core modules." -T
+        $ProgressPreference = $SavedProgressPreference
+    }
+
+    RunUpdateHelpTests -tag "Feature" -Scope 'CurrentUser'
```
@adityapatwardhan @SteveL-MSFT  I wonder why we use `-Scope` if the function have only `[switch]$userscope`?

adityapatwardhan replied:
From a quick look seems to be a bug in the test.
